File: internal_auth.py
"""
Internal Authentication Middleware for SMSLYCLOUD Microservices

This middleware validates SPIFFE mTLS identity for gateway-to-service
communication. Should be added to all FastAPI microservices.

Phase 4: HMAC removed. Only SPIFFE mTLS accepted.

Usage:
    from internal_auth import InternalAuthMiddleware, get_internal_context
    
    app.add_middleware(InternalAuthMiddleware, service_name="my-service")
    
    @app.get("/v1/messages")
    async def list_messages(context: InternalContext = Depends(get_internal_context)):
        user_id = context.user_id
        ...
"""

from dataclasses import dataclass
from typing import Optional
from fastapi import Request, HTTPException
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import logging

try:
    from smsly_core.spiffe_auth import (
        DualAuthValidator,
        get_allowed_callers,
        MigrationPhase,
    )
    _SPIFFE_AVAILABLE = True
except ImportError:
    _SPIFFE_AVAILABLE = False

logger = logging.getLogger(__name__)


@dataclass
class InternalContext:
    """Context passed from the gateway to the microservice."""
    user_id: Optional[str] = None
    user_email: Optional[str] = None
    organization_id: Optional[str] = None
    account_type: str = "casual"
    request_id: Optional[str] = None
    is_internal: bool = False
    spiffe_id: Optional[str] = None


class InternalAuthMiddleware(BaseHTTPMiddleware):
    """
    Middleware to validate internal authentication from the gateway.
    
    Validates SPIFFE mTLS identity and extracts user context.
    
    Phase 4: HMAC removed. Only SPIFFE mTLS accepted.
    """
    
    def __init__(self, app, skip_paths: list = None, service_name: str = "unknown"):
        super().__init__(app)
        self.skip_paths = skip_paths or ["/health", "/docs", "/redoc", "/openapi.json", "/metrics"]
        self.service_name = service_name

        # Initialize SPIFFE validator
        self._spiffe_validator = None
        if _SPIFFE_AVAILABLE:
            allowed = get_allowed_callers(service_name)
            self._spiffe_validator = DualAuthValidator(
                service_name=service_name,
                allowed_callers=allowed,
            )
            logger.info(
                "internal_auth_initialized",
                service=service_name,
                phase=self._spiffe_validator.flags.migration_phase.value,
            )
        else:
            logger.error(
                "internal_auth_no_spiffe",
                service=service_name,
                msg="SPIFFE module not available — all requests will be rejected",
            )

    async def dispatch(self, request: Request, call_next):
        # Skip health checks and documentation
        if any(request.url.path.startswith(p) for p in self.skip_paths):
            return await call_next(request)

        # SPIFFE mTLS validation
        if not self._spiffe_validator:
            logger.error("internal_auth_no_validator", path=request.url.path)
            return JSONResponse(
                status_code=503,
                content={"error": "authentication_unavailable", "detail": "Auth service not configured"}
            )

        headers = {k: v for k, v in request.headers.items()}
        result = self._spiffe_validator.validate(
            headers=headers,
            method=request.method,
            path=request.url.path,
        )

        if not result.authenticated:
            logger.warning(
                "internal_auth_failed",
                path=request.url.path,
                reason=result.reason,
                client=request.client.host if request.client else "unknown",
            )
            return JSONResponse(
                status_code=401,
                content={
                    "error": "Unauthorized",
                    "detail": result.reason or "Valid SPIFFE mTLS identity required",
                }
            )

        # Extract context from headers
        context = InternalContext(
            user_id=request.headers.get("X-User-ID"),
            user_email=request.headers.get("X-User-Email"),
            organization_id=request.headers.get("X-Organization-ID"),
            account_type=request.headers.get("X-Account-Type", "casual"),
            request_id=request.headers.get("X-Request-ID"),
            is_internal=True,
            spiffe_id=result.spiffe_id,
        )

        # Store context in request state
        request.state.internal_context = context

        # Add request ID to response headers
        response = await call_next(request)
        if context.request_id:
            response.headers["X-Request-ID"] = context.request_id

        return response


def get_internal_context(request: Request) -> InternalContext:
    """
    Dependency to get internal context from request.
    
    Usage:
        @app.get("/v1/resource")
        async def get_resource(context: InternalContext = Depends(get_internal_context)):
            print(context.user_id)
    """
    context = getattr(request.state, 'internal_context', None)
    if context is None:
        # Create empty context for non-gateway requests (e.g., direct API access)
        context = InternalContext(is_internal=False)
    return context


def require_internal_auth(request: Request) -> InternalContext:
    """
    Dependency that requires internal authentication.
    Raises 401 if request is not from gateway.
    """
    context = get_internal_context(request)
    if not context.is_internal:
        raise HTTPException(
            status_code=401,
            detail="This endpoint requires internal gateway authentication"
        )
    return context


def require_user_context(request: Request) -> InternalContext:
    """
    Dependency that requires user context from gateway.
    Raises 401 if no user context is present.
    """
    context = require_internal_auth(request)
    if not context.user_id:
        raise HTTPException(
            status_code=401,
            detail="User context required"
        )
    return context


# Rate limiter that respects account type
class AccountTypeRateLimiter:
    """
    Rate limiter that uses different limits based on account type.
    """
    
    DEFAULT_LIMITS = {
        "casual": {"requests_per_second": 5, "requests_per_minute": 60},
        "developer": {"requests_per_second": 20, "requests_per_minute": 300},
        "enterprise": {"requests_per_second": 100, "requests_per_minute": 1000},
        "reseller": {"requests_per_second": 50, "requests_per_minute": 500},
    }
    
    def __init__(self, redis_client=None, limits: dict = None):
        self.redis = redis_client
        self.limits = limits or self.DEFAULT_LIMITS
    
    async def check_rate_limit(self, context: InternalContext) -> bool:
        """Check if request is within rate limits."""
        if self.redis is None:
            logger.warning("rate_limit_no_redis", msg="Redis unavailable, denying request")
            return False  # Fail-closed
        
        account_type = context.account_type
        limits = self.limits.get(account_type, self.limits["casual"])
        
        # Use user_id or organization_id as key
        key_base = context.organization_id or context.user_id or "anonymous"
        
        # Check per-second limit
        second_key = f"rate:{key_base}:second"
        current = await self.redis.incr(second_key)
        if current == 1:
            await self.redis.expire(second_key, 1)
        if current > limits["requests_per_second"]:
            return False
        
        # Check per-minute limit
        minute_key = f"rate:{key_base}:minute"
        current = await self.redis.incr(minute_key)
        if current == 1:
            await self.redis.expire(minute_key, 60)
        if current > limits["requests_per_minute"]:
            return False
        
        return True

File: smsly-rust/smsly-services/src/internal_auth.rs
use crate::config::Settings;
use axum::{
    extract::{Request, State},
    http::StatusCode,
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use constant_time_eq::constant_time_eq;
use redis::{AsyncCommands, Client, Script};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::sync::Arc;
use tracing::warn;

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct InternalContext {
    pub user_id: Option<String>,
    pub user_email: Option<String>,
    pub organization_id: Option<String>,
    #[serde(default = "default_account_type")]
    pub account_type: String,
    pub request_id: Option<String>,
    pub is_internal: bool,
}

fn default_account_type() -> String {
    "casual".to_string()
}

#[derive(Clone)]
pub struct AppState {
    pub redis: Option<Client>,
    pub settings: Settings,
}

pub async fn internal_auth_middleware(
    State(state): State<Arc<AppState>>,
    mut request: Request,
    next: Next,
) -> Result<Response, StatusCode> {
    let path = request.uri().path();
    let skip_paths = ["/health", "/docs", "/redoc", "/openapi.json", "/metrics"];
    if skip_paths.iter().any(|p| path.starts_with(p)) {
        return Ok(next.run(request).await);
    }

    let headers = request.headers();
    let provided_secret = headers
        .get("X-Internal-Secret")
        .and_then(|h| h.to_str().ok())
        .unwrap_or("");

    let internal_secret = &state.settings.internal_api_secret;

    if internal_secret.is_empty() {
        warn!("INTERNAL_API_SECRET not configured, allowing all requests");
    } else if !constant_time_eq(provided_secret.as_bytes(), internal_secret.as_bytes()) {
        warn!("Invalid internal secret from {:?}", request.uri());
        return Ok((
            StatusCode::UNAUTHORIZED,
            Json(json!({"error": "Unauthorized", "detail": "Invalid internal secret"})),
        )
            .into_response());
    }

    let context = InternalContext {
        user_id: headers
            .get("X-User-ID")
            .and_then(|h| h.to_str().ok())
            .map(String::from),
        user_email: headers
            .get("X-User-Email")
            .and_then(|h| h.to_str().ok())
            .map(String::from),
        organization_id: headers
            .get("X-Organization-ID")
            .and_then(|h| h.to_str().ok())
            .map(String::from),
        account_type: headers
            .get("X-Account-Type")
            .and_then(|h| h.to_str().ok())
            .unwrap_or("casual")
            .to_string(),
        request_id: headers
            .get("X-Request-ID")
            .and_then(|h| h.to_str().ok())
            .map(String::from),
        is_internal: true,
    };

    if let Some(redis_client) = &state.redis {
        let limiter = AccountTypeRateLimiter::new(redis_client.clone());
        if !limiter.check_rate_limit(&context).await {
            return Ok((
                StatusCode::TOO_MANY_REQUESTS,
                Json(json!({"error": "Too Many Requests", "detail": "Rate limit exceeded"})),
            )
                .into_response());
        }
    }

    request.extensions_mut().insert(context.clone());

    let mut response = next.run(request).await;

    if let Some(req_id) = context.request_id {
        response
            .headers_mut()
            .insert("X-Request-ID", req_id.parse().unwrap());
    }

    Ok(response)
}

pub struct AccountTypeRateLimiter {
    redis: Client,
}

impl AccountTypeRateLimiter {
    pub fn new(redis: Client) -> Self {
        Self { redis }
    }

    pub async fn check_rate_limit(&self, context: &InternalContext) -> bool {
        let key_base = context
            .organization_id
            .as_deref()
            .or(context.user_id.as_deref())
            .unwrap_or("anonymous");

        let (limit_sec, limit_min) = match context.account_type.as_str() {
            "developer" => (20, 300),
            "enterprise" => (100, 1000),
            "reseller" => (50, 500),
            _ => (5, 60),
        };

        let mut conn = match self.redis.get_multiplexed_async_connection().await {
            Ok(c) => c,
            Err(e) => {
                warn!("Redis connection failed for rate limit: {}", e);
                return true;
            }
        };

        let script = Script::new(
            r#"
            local current = redis.call("INCR", KEYS[1])
            if current == 1 then
                redis.call("EXPIRE", KEYS[1], ARGV[1])
            end
            return current
        "#,
        );

        let second_key = format!("rate:{}:second", key_base);
        let current_sec: u64 = match script.key(&second_key).arg(1).invoke_async(&mut conn).await {
            Ok(v) => v,
            Err(_) => return true,
        };
        if current_sec > limit_sec {
            return false;
        }

        let minute_key = format!("rate:{}:minute", key_base);
        let current_min: u64 = match script
            .key(&minute_key)
            .arg(60)
            .invoke_async(&mut conn)
            .await
        {
            Ok(v) => v,
            Err(_) => return true,
        };
        if current_min > limit_min {
            return false;
        }

        true
    }
}
